### PR TITLE
Eliminate race in connection closed tcp test case

### PR DESCRIPTION
**Pull request by johanandren**
Refs #21903

I have changed the test so that it should not race anymore, but either I'm wrong about that or there is a race-bug in the tcp server stuffs because the test still fails sporadically.

### CODE REVIEW

File: akka-stream-tests/src/test/scala/akka/stream/io/TcpSpec.scala [hunk 1]
```diff
@@ -508,11 +508,30 @@ class TcpSpec extends StreamSpec("akka.stream.materializer.subscription-timeout.
...
+        } catch {
+          case _: SocketException ⇒ true
+        }
+
+      import system.dispatcher
+      val futureBinding: Future[ServerBinding] =
+        Tcp().bind(address.getHostName, address.getPort)
+          // accept one connection, then cancel
+          .take(1)
+          // keep the accepted request hanging, but do not block the
```
Comment (2m) [inline akka-stream-tests/src/test/scala/akka/stream/io/TcpSpec.scala]: I like the cliffhanger in the comments. ;)

**Comment by johanandren** [comment]
With TCP trace logging on, I can see that every time that it fails, there is a `setSoLinger(true, 0) failed with [java.net.SocketException: Invalid argument]` logged, which in the code has a comment referring to http://bugs.sun.com/view_bug.do?bug_id=6799574

**Comment by johanandren** [comment]
Amended with the complete test I'm running now, attached is a failure output with tcp trace on.

[failure.txt](https://github.com/akka/akka/files/619660/failure.txt)

**Comment by ktoso** [comment]
The test looks fine from the outside... I'll try playing with it locally,

**Comment by patriknw** [comment]
PLS BUILD

**Comment by johanandren** [comment]
Do not merge this PR, it's very unstable (possibly actually exercising a bug)

**Comment by ktoso** [comment]
Ok, please put into title then 👍

**Comment by drewhk** [comment]
Should this be reviewed or is this in-progress? If in-progress I propose closing it and reopening when it is ready to review.

**Comment by johanandren** [comment]
@drewhk I got to a state where I think I have fixed the tests, but that manifests in more frequent failures, which I could not figure out if it is a bug show its face or if I actually rather broke the test than fixing it.

**Comment by johanandren** [comment]
`setSoLinger` was probably a dead end. Now I have observed the following, the `ConnectionSourceStage` sends `Unbind` to the listener, but 10 seconds later it has still not gotten a `Unbound` message back, even though I can see that the listener has unbound endpoint and is stopping itself. O.o

**Comment by drewhk** [comment]
Maybe some state transition where the message is not sent? Btw, shouldn't the Source stage watch the listener actor anyway?

**Comment by johanandren** [comment]
It does watch the Listener, but when it tries to unbind it unwatches it (since it will stop itself I guess). Not quite sure that has anything to do with the failure yet though.

**Comment by drewhk** [comment]
hm, shouldn't it unwatch after the unbind was successful? I mean if the actor goes away silently we wouldn't see.

File: akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala [hunk 2]
```diff
@@ -84,11 +85,14 @@ private[stream] class ConnectionSourceStage(
...
-            if (connectionFlowsAwaitingInitialization.get() == 0) completeStage()
-            else scheduleOnce(BindShutdownTimer, bindShutdownTimeout)
+          case Unbound ⇒
+            unbindCompleted()
           case Terminated(ref) if ref == listener ⇒
-            failStage(new IllegalStateException("IO Listener actor terminated unexpectedly"))
+            if (unbindStarted) {
+              unbindCompleted()
+            } else {
+              failStage(new IllegalStateException("IO Listener actor terminated unexpectedly"))
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala]: include `endpoint` in exception message?

**Review by ktoso — APPROVED** [review]
LGTM... I guess

**Review by patriknw — APPROVED** [review]
LGTM, I guess

**Comment by johanandren** [comment]
I'll merge then, I guess

**Pull request merged**